### ISSUE TITLE
ASN Statistics broken

### ISSUE THREAD

**Issue by vidister** [monogon-dev/NetMeta#72]
In our setups with exports from Junos based routers (using both sflow and netflow) the default dashboard doesn't display any ASN statistics.
When I remove the `AND FlowDirection = ` conditions from the queries everything works fine.
Maybe our routers don't add FlowDirection Attributes to the flowsamples?
Is it possible to remove the condition?

**Comment by vidister** [comment]
> Happy to implement this - if I understood the linked article correctly, this would result in correct-ish data by counting all ingress/egress traffic in both tables, right?

this is correct.

> We already do something similar for the other graphs: if(FlowDirection == 1, 'out', 'in') AS FlowDirection

Yes, in that case every flow is labeled as "in" right now, which can be a bit confusing.
What are we doing with these? We could put a third string "Unknown" in there, but I think that could also be confusing in the dashboard. Maybe leaving the string empty could do the job? Still a bit messy when mixed sources with differently broken flow export implementations are involved, but probably as good as it gets...

**Comment by leoluk** [comment]
Sounds to me like the "correct" solution is to fix up the data at ingestion time. Your data seems to have correct `InIf` and `OutIf` data, so presumably one could deduce the correct flow direction from that?

**Comment by vidister** [comment]
But how do you know if a interface is a edge-port or some internal/backbone interface? Set a flag in the interfaceMap?

**Comment by leoluk** [comment]
Yup, that was the idea - just have a map of all interfaces and which way they're facing, possibly determined from Netbox and/or SNMP. Does that sounds workable?

**Comment by vidister** [comment]
Not nice, but probably the best solution.

**Comment by leoluk** [comment]
Yeah... I don't think we can avoid it unless the device tells us the physical flow direction, which it doesn't want to...

How about having a list of "local" CIDR ranges?

**Comment by vidister** [comment]
This would work fine for Hosting-Provider-Like networks but not so well for ISP networks with downstream ASNs (like ours.).
I think fastnetmon solves this by receiving the routes via BGP... which seems a bit overkill here.

**Comment by leoluk** [comment]
Hmm...we could do that! No half measures 😆

How would that look like - use BGP/BMP to figure out local networks?

**Comment by vidister** [comment]
Yup. BGP/BMP integration could become useful anyways. We could work with BGP communities.

**Comment by leoluk** [comment]
Okay, let's do that - sounds like the "correct" solution.

**Comment by vidister** [comment]
> i.e. have a config setting which communities mark "local" networks?

yes. And we can generalize this to use other communities as well for filtering.. For example customer networks, region/city communities, etc.

**Comment by fionera** [comment]
The ASN Lookup is now fixed on the development branch by #87 / #88 .  Can the issue then be closed or do we still have the FlowDirection issue?

**Comment by leoluk** [comment]
This is now implemented, thanks again :)

**Comment by fionera** [comment]
The FlowDirection issue is still open :/

**Comment by fionera** [comment]
The issue with the FlowDirection is partly happening on portmirror deployments too, because the flow data that gets ingested only has one InIf/OutIf set. Because of this, all Graphs have the sum of all traffic from the other direction displayed. We should probably find a way to infer it ourselves since it is the broader solution to issues like this
